File: Python/113_Path Sum II.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:         
    def pathSum(self, root: TreeNode, targetSum):
        ans = []  
        def backtracking(node,path,path_res):
            if not node:
                return             
            path +=[node.val]  
            path_res -= node.val         
            if not node.left and not node.right and path_res == 0:
                ans.append(path[:])
            else:
                backtracking(node.left,path,path_res)
                backtracking(node.right,path,path_res)                 
            path.pop(-1)



        backtracking(root,[],targetSum)
        return ans        
            

File: C#/113.path-sum-ii.cs
/*
 * @lc app=leetcode id=113 lang=csharp
 *
 * [113] Path Sum II
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
 using System.Collections.Generic;
 public class Solution {
        public IList<IList<int>> PathSum(TreeNode root, int targetSum)
        {
            IList<IList<int>> ans = new List<IList<int>>();            
            getPathSum(root, targetSum, new List<int>(), ans);
            return ans;
        }

        public void getPathSum(TreeNode root, int path_res, IList<int> path, IList<IList<int>> ans)
        {                        
            if (node == null)
                return;

            path.Add(node.val);
            path_res = path_res - node.val;               
            getPathSum(node.left, path_res, new List<int>(path), ans);  
            getPathSum(node.right, path_res, new List<int>(path), ans);  
            if (node.left == null && node.right == null && path_res == 0)           
                ans.Add(path);
            

        }

}
// @lc code=end

